src/renderer/base/FontResource.cpp

```diff
@@ -0,0 +1,285 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+#include "precomp.h"
+#include "../inc/FontResource.hpp"
+
+using namespace Microsoft::Console::Render;
+
+namespace
+{
+    // The structures below are based on the Windows 3.0 font file format, which
+    // was documented in Microsoft Knowledge Base article Q65123. Although no
+    // longer hosted by Microsoft, it can still be found at the following URL:
+    // https://web.archive.org/web/20140820153410/http://support.microsoft.com/kb/65123
+
+    // For now we're only using fixed pitch single color fonts, but the rest
+    // of the flags are included here for completeness.
+    static constexpr DWORD DFF_FIXED = 0x0001;
+    static constexpr DWORD DFF_PROPORTIONAL = 0x0002;
+    static constexpr DWORD DFF_1COLOR = 0x0010;
+    static constexpr DWORD DFF_16COLOR = 0x0020;
+    static constexpr DWORD DFF_256COLOR = 0x0040;
+    static constexpr DWORD DFF_RGBCOLOR = 0x0080;
+
+    // DRCS soft fonts only require 96 characters at most.
+    static constexpr size_t CHAR_COUNT = 96;
+
+#pragma pack(push, 1)
+    struct GLYPHENTRY
+    {
+        WORD geWidth;
+        DWORD geOffset;
+    };
+
+    struct FONTINFO
+    {
+        WORD dfVersion;
+        DWORD dfSize;
+        CHAR dfCopyright[60];
+        WORD dfType;
+        WORD dfPoints;
+        WORD dfVertRes;
+        WORD dfHorizRes;
+        WORD dfAscent;
+        WORD dfInternalLeading;
+        WORD dfExternalLeading;
+        BYTE dfItalic;
+        BYTE dfUnderline;
+        BYTE dfStrikeOut;
+        WORD dfWeight;
+        BYTE dfCharSet;
+        WORD dfPixWidth;
+        WORD dfPixHeight;
+        BYTE dfPitchAndFamily;
+        WORD dfAvgWidth;
+        WORD dfMaxWidth;
+        BYTE dfFirstChar;
+        BYTE dfLastChar;
+        BYTE dfDefaultChar;
+        BYTE dfBreakChar;
+        WORD dfWidthBytes;
+        DWORD dfDevice;
+        DWORD dfFace;
+        DWORD dfBitsPointer;
+        DWORD dfBitsOffset;
+        BYTE dfReserved;
+        DWORD dfFlags;
+        WORD dfAspace;
+        WORD dfBspace;
+        WORD dfCspace;
+        DWORD dfColorPointer;
+        DWORD dfReserved1[4];
+        GLYPHENTRY dfCharTable[CHAR_COUNT];
+        CHAR szFaceName[LF_FACESIZE];
+    };
+#pragma pack(pop)
+}
+
+FontResource::FontResource(const gsl::span<const uint16_t> bitPattern,
+                           const til::size sourceSize,
+                           const til::size targetSize,
+                           const size_t centeringHint) :
+    _bitPattern{ bitPattern.begin(), bitPattern.end() },
+    _sourceSize{ sourceSize },
+    _targetSize{ targetSize },
+    _centeringHint{ centeringHint }
+{
+}
+
+void FontResource::SetTargetSize(const til::size targetSize)
+{
+    if (_targetSize != targetSize)
+    {
+        _targetSize = targetSize;
+        _fontHandle = nullptr;
+    }
+}
+
+FontResource::operator HFONT()

```

Comment:
I was kind of hoping we could just register the font the same way we are in GDI with `AddFontMemResourceEx`, and the given font name would then be available for use in DX, but maybe that's nonsense. And that's still dependent on a GDI API so in that respect it's no better than option 2.

Otherwise if we have to create a TTF or OTF font, I was wondering whether there might still be a way to do that with a raster format. I came across something called embedded bitmap data in the OpenType format, which sounded promising (see https://docs.microsoft.com/en-us/typography/opentype/otspec150/eblc). But again I don't know anything about the subject, so maybe that's not what I think it is.

And as a last resort, there's always the possibility of creating a texture with all the characters, and then blitting the appropriate sections of the texture for each run of characters.

Reply:
Yeah what you're saying with `AddFontMemResourceEx` would be the precursor to doing all the other `CreateFontFaceFromHdc` stuff... so it's just a part of number 2.

I think @DHowett mentioned to me the other day that DirectWrite doesn't look at the embedded bitmap data in TTF/OTF... or something like that. We'd have to likely talk to the DirectWrite team about it...

The last resort might honestly be the easiest one, unfortunately.

Reply:
> > there's always the possibility of creating a texture with all the characters, and then blitting the appropriate sections of the texture for each run of characters
>
> The last resort might honestly be the easiest one

lol

Reply:
(For James' edification--though I suspect he is entirely aware--between Michael's comment in May and Mike's yesterday it became obvious that an atlas-based renderer would be terrifically effective for us, and we've put some time into prototyping one that will eventually ship.)

Reply:
If you're interested you can check out https://github.com/microsoft/terminal/compare/dev/lhecker/atlas-engine for this. I know the branch is called "atlas-engine", but it actually contains a mishmash of various branches I built over time. The final PR I'm submitting will only contain the AtlasEngine project.
If anyone wants to try it out: It's only integrated in `Host.EXE` and can be enabled by setting `UseDx` to 2 under `HKCU:\Console` (`UseDx == 1` is the current DxEngine).